### ISSUE TITLE
CS8600 with MemberNotNullWhen in a property and using an implicit operator.

### ISSUE THREAD

**Issue by jeroenlandheer** [dotnet/roslyn#66160]
**Brief description:**

See the example code, the CS8600 shouldn't be there.

**Languages applicable:**

C#, and maybe others?

**Code example that the analyzer should report:**

```csharp
namespace TestNullApp
{
    using System.Diagnostics.CodeAnalysis;

    class Program
    {
        static void Main(string[] args)
        {
            var testNull = new TestNull("Hello");
            if (!testNull.IsNull)
            {
                string test = testNull; // CS8600
                Console.WriteLine(test);
            }
        }
    }

    internal readonly struct TestNull
    {
        private readonly string? value;

        public TestNull(string? value)
        {
            this.value = value;
        }

        [MemberNotNullWhen(false, nameof(value))]
        public bool IsNull => this.value == null;

        public static implicit operator string?(TestNull testNull)
        {
            return testNull.value;
        }
    }
}

```

**Comment by svick** [comment]
Nullable analysis does not see inside methods, so it does not understand the relationship between the `value` field and the implicit operator. So I think this behaves as designed and it's very unlikely to change.

**Comment by jeroenlandheer** [comment]
I found this issue with code that was using StackExchange.Redis (a popular Redis client), but probably other libraries might also have similar issues; I don't think this scenario is an edge-case.  If this analyser's accuracy can be improved this way, it might also help with other code. 

I can imagine this won't be easy though 🤔

**Comment by CyrusNajmabadi** [comment]
There's no way for the compiler to know what the contents of a method are. Your operator says it returns a `string?` and the compiler believes that. 

Here, you can change the API to expose that value through a property which you have the appropriate NRT attributes explain, and things will work correctly.

**Comment by CyrusNajmabadi** [comment]
>  I don't think this scenario is an edge-case

This is the first report :-)

It's not a goal to support every pattern of every library. There would need to be a ton of usage of this and a lot of feedback that this was a problem to create a customized solution for this case

**Comment by jaredpar** [comment]
> Nullable analysis does not see inside methods, so it does not understand the relationship between the value field and the implicit operator. So I think this behaves as designed and it's very unlikely to change.

This is the case. The nullable analysis, nor really any analysis in the compiler, considers the content of method bodies. This is not going to change anytime soon.
